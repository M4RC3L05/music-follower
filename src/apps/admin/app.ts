import config from "config";
import Koa from "koa";
import bodyParser from "koa-bodyparser";
import render from "koa-ejs";
import qs from "koa-qs";
import session from "koa-session";
import koaStatic from "koa-static";
import path from "node:path";
import url from "node:url";

import router from "#src/apps/admin/router.js";

export const app = () => {
  const koa = new Koa();
  koa.keys = config.get("apps.admin.keys");

  qs(koa);
  render(koa, {
    root: path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), "views"),
    layout: "main",
    viewExt: "ejs",
    cache: process.env.NODE_ENV === "production",
    debug: false,
  });

  koa.use(session({ ...config.get("apps.admin.session") }, koa));
  koa.use(bodyParser());
  koa.use(koaStatic(path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), "../../../public")));
  koa.use(router.middleware());
  koa.use(router.allowedMethods());

  return koa;
};
