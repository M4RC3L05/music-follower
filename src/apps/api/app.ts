import { type Database } from "@leafac/sqlite";
import config from "config";
import { Hono } from "hono";
import { basicAuth } from "hono/basic-auth";
import { cors } from "hono/cors";
import { errorMappers } from "#src/errors/mod.js";
import { errorMapper, requestLifeCycle } from "#src/middlewares/mod.js";
import { router } from "./router.js";

export type Api = ReturnType<typeof makeApp>;

export const makeApp = ({ database }: { database: Database }) => {
  const app = new Hono();

  app.use("*", (c, next) => {
    c.set("database", database);

    return next();
  });
  app.use("*", requestLifeCycle);
  app.use("*", cors());
  app.use(
    "*",
    basicAuth({
      username: config.get<{ name: string; pass: string }>("apps.api.basicAuth")
        .name,
      password: config.get<{ name: string; pass: string }>("apps.api.basicAuth")
        .pass,
    }),
  );

  app.onError(
    errorMapper({
      mappers: [errorMappers.validationErrorMapper],
      defaultMapper: errorMappers.defaultErrorMapper,
    }),
  );

  return app.route("/api", router());
};
