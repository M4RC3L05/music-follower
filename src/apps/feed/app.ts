import { App } from "@m4rc3l05/sss";

import { feedMiddleware } from "./middlewares/feed-middleware.js";
import requestLifeCycle from "#src/middlewares/request-lifecycle.js";

export const makeApp = () => {
  const app = new App();

  app.use(requestLifeCycle);
  app.use(feedMiddleware);

  return app;
};
