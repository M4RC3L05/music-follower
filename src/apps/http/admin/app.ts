import Koa from "koa";
import basicAuth from "koa-basic-auth";
import config from "config";
import koaStatic from "koa-static";

export const app = () => {
  const koa = new Koa();

  koa.use(basicAuth({ ...config.get("apps.admin.basicAuth") }));
  koa.use(koaStatic("./src/apps/http/admin/public"));

  return koa;
};
