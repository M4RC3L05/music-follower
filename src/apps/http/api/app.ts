import Koa from "koa";
import bodyParser from "koa-bodyparser";
import cors from "@koa/cors";

import * as errorMappers from "#src/common/errors/mappers/mod.js";
import { errorMapper, requestLifeCycle } from "#src/common/middlewares/mod.js";
import logger from "#src/common/clients/logger.js";
import router from "./router.js";

export const app = () => {
  const app = new Koa();

  app.use(requestLifeCycle({ loggerFactory: logger }));
  app.use(
    errorMapper({
      loggerFactory: logger,
      defaultMapper: errorMappers.defaultErrorMapper,
      mappers: [errorMappers.validationErrorMapper],
    }),
  );
  app.use(cors());
  // App.use(basicAuth({ ...config.get("apps.api.basicAuth") }));
  app.use(bodyParser());
  app.use(router.routes());
  app.use(router.allowedMethods());

  return app;
};
