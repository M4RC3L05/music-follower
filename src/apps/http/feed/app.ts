import Koa from "koa";
import favicon from "koa-favicon";

import { feedMiddleware } from "#src/apps/http/feed/middlewares/feed-middleware.js";

export const app = () => {
  const koa = new Koa();

  koa.use(favicon("./src/apps/http/feed/favicon.ico"));
  koa.use(feedMiddleware);

  return koa;
};
