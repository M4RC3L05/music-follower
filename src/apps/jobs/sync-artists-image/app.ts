import { makeLogger } from "#src/common/logger/mod.ts";
import type { Artist } from "#src/database/types/mod.ts";
import type { CustomDatabase } from "#src/database/mod.ts";
import { delay } from "@std/async";
import { getArtistById } from "#src/remote/itunes/requests/mod.ts";
import { getArtistImage } from "#src/remote/apple-music/requests/mod.ts";

const log = makeLogger("sync-artists-image-job");

const runner = async ({
  db,
  abort,
}: { db: CustomDatabase; abort: AbortSignal }) => {
  if (abort.aborted) {
    return;
  }

  log.info("Begin artist image sync");

  const artists = db.prepare(`
    select *, row_number() over (order by id) as "index", ti."totalItems" as "totalItems"
    from
      artists,
      (select count(id) as "totalItems" from artists) as ti
    order by id;
  `).iter() as IterableIterator<Artist & { index: number; totalItems: number }>;

  for (const artist of artists) {
    const isLastArtist = artist.index >= artist.totalItems;

    if (abort.aborted) {
      break;
    }

    log.info(
      `Sync image from "${artist.name}" at ${artist.index} of ${artist.totalItems}`,
    );

    let imageUrl: string;

    try {
      const artistRemote = await getArtistById(artist.id, abort);

      if (!artistRemote) {
        throw new Error(`No remote artist found with id ${artist.id}`);
      }

      imageUrl = await getArtistImage(
        artistRemote.artistLinkUrl,
        abort,
      );
    } catch (error: unknown) {
      log.error("Something went wrong fetching artist image", { error });

      if (!isLastArtist) {
        log.info("Waiting 5 seconds before processing next artist");

        await delay(5000, { signal: abort }).catch(() => {});
      }

      continue;
    }

    if (abort.aborted) {
      break;
    }

    try {
      log.info("Updating artist image", { id: artist.id });

      db.sql`update artists set image = ${imageUrl} where id = ${artist.id}`;

      log.info("Artist image updated", { id: artist.id });
    } catch (error: unknown) {
      log.error("Something wrong ocurred while updating artist image", {
        error,
      });
    }

    if (!isLastArtist) {
      log.info("Waiting 5 seconds before processing next artist");

      await delay(5000, { signal: abort }).catch(() => {});
    }
  }

  log.info("Artist image sync ended");
};

export default runner;
