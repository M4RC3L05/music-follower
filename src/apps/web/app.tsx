import { serveStatic } from "@hono/hono/deno";
import config from "config";
import { type ContextVariableMap, Hono } from "@hono/hono";
import {
  MemoryStore,
  type Session,
  sessionMiddleware,
} from "@jcs224/hono-sessions";
import { csrf } from "@hono/hono/csrf";
import { jsxRenderer } from "@hono/hono/jsx-renderer";
import { HTTPException } from "@hono/hono/http-exception";
import { secureHeaders } from "@hono/hono/secure-headers";
import { makeLogger } from "#src/common/logger/mod.ts";
import { router } from "#src/apps/web/routes/mod.ts";
import { MainLayout } from "#src/apps/web/views/common/layouts/main.tsx";
import { serviceRegister } from "#src/middlewares/mod.ts";
import type { CustomDatabase } from "#src/database/mod.ts";
import { PagesErrorPage } from "./views/pages/pages/error.tsx";
import type { MiddlewareHandler } from "@hono/hono";
import { STATUS_TEXT, type StatusCode } from "@std/http";
import type {
  SessionFlashFormErrors,
  SessionFlashMessages,
} from "#src/apps/web/types.ts";
import { MultipartParseError } from "@mjackson/multipart-parser";
import { errors } from "@vinejs/vine";
import { feedRoutes } from "#src/apps/web/routes/feed/mod.ts";

const log = makeLogger("admin");

declare module "@hono/hono" {
  interface ContextVariableMap {
    database: CustomDatabase;
    shutdown: AbortSignal;
    session: Session<
      {
        flashMessages?: SessionFlashMessages;
        flashFormErrors?: SessionFlashFormErrors;
      }
    >;
  }

  interface ContextRenderer {
    (
      content: string | Promise<string>,
      props?: { error?: boolean },
    ): Response;
  }
}

export const makeApp = (deps: Partial<ContextVariableMap>) => {
  const app = new Hono();

  app.onError((error, c) => {
    log.error("Something went wrong!", { error });

    if (error instanceof HTTPException) {
      if (error.res) {
        error.res.headers.forEach((value, key) => {
          c.header(key, value);
        });
      }
    }

    if ((error instanceof HTTPException && error.status === 422)) {
      c.get("session").flash("flashMessages", { error: [error.message] });

      if (error.cause) {
        c.get("session").flash(
          "flashFormErrors",
          error.cause as SessionFlashFormErrors,
        );
      }

      return c.redirect(c.req.header("Referer") ?? "/");
    }

    if (error instanceof errors.E_VALIDATION_ERROR) {
      c.get("session").flash("flashMessages", { error: [error.message] });
      c.get("session").flash(
        "flashFormErrors",
        (error.messages as { message: string; field: string }[]).reduce(
          (acc, curr) => {
            const errMessages = acc[curr.field] ?? [];
            errMessages.push(curr.message);
            acc[curr.field] = errMessages;
            return acc;
          },
          {} as Record<string, string[]>,
        ),
      );

      return c.redirect(c.req.header("Referer") ?? "/");
    }

    if (error instanceof MultipartParseError) {
      c.get("session").flash("flashMessages", { error: [error.message] });
      return c.redirect(c.req.header("Referer") ?? "/");
    }

    const errorMessage = error instanceof HTTPException
      ? (error.message.trim().length > 0 ? error.message : undefined) ??
        STATUS_TEXT[error.status as StatusCode] ??
        "Something broke..."
      : "Something broke...";
    c.status(error instanceof HTTPException ? error.status : 500);

    return c.render(<PagesErrorPage errorMessage={errorMessage} />, {
      error: true,
    });
  });

  app.notFound((c) => {
    c.status(404);
    return c.render(<PagesErrorPage errorMessage="Not found" />, {
      error: true,
    });
  });

  app.use("*", async (c, next) => {
    const nocache = () => {
      // This is important so that we always make sure the browser will not cache the previous page
      // so that the requests are always made.
      c.header("cache-control", "no-cache, no-store, must-revalidate");
      c.header("pragma", "no-cache");
      c.header("expires", "0");
    };

    try {
      await next();

      if (
        !(c.req.path.includes("/favicon") || c.req.path.includes("/public") ||
          c.req.path.includes("/deps"))
      ) {
        nocache();
      } else {
        c.header("Cache-Control", `max-age=${60 * 60 * 24}`);
      }
    } catch (error) {
      if (
        !(c.req.path.includes("/favicon") || c.req.path.includes("/public") ||
          c.req.path.includes("/deps"))
      ) {
        nocache();
      } else {
        c.header("Cache-Control", `max-age=${60 * 60 * 24}`);
      }

      throw error;
    }
  });
  app.use("*", serviceRegister(deps));
  app.use(
    "*",
    secureHeaders({ referrerPolicy: "strict-origin-when-cross-origin" }),
  );

  app.route("/feeds", feedRoutes());
  app.get("/favicon.ico", serveStatic({ root: "./src/apps/web/public" }));
  app.get(
    "/public/*",
    serveStatic({
      root: "./src/apps/web/public",
      rewriteRequestPath: (path) => path.replace("/public", ""),
    }),
  );
  app.get(
    "/deps/*",
    serveStatic({
      root: "./node_modules",
      rewriteRequestPath: (path) => path.replace("/deps", ""),
    }),
  );

  app.use(
    "*",
    sessionMiddleware({
      store: new MemoryStore(),
      encryptionKey: "foobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobar",
      expireAfterSeconds: 60 * 15,
      sessionCookieName: "sid",
      cookieOptions: {
        sameSite: "strict",
        path: "/",
        httpOnly: true,
        secure: true,
      },
    }) as unknown as MiddlewareHandler,
  );

  app.use(
    "*",
    jsxRenderer(({ children, error = false }, c) => {
      const flashMessages = c.get("session").get("flashMessages");

      return (
        <MainLayout
          path={c.req.path}
          flashMessages={flashMessages ?? undefined}
          enableNavbar={!error}
        >
          {children}
        </MainLayout>
      );
    }, {
      docType: true,
      stream: true,
    }),
  );

  app.use(
    "*",
    csrf({
      origin: (origin) => new RegExp(config.get("csrf.origin")).test(origin),
    }),
  );

  if (Deno.env.get("ENV") === "test") {
    app.get("/test/flash-messages", (c) => {
      c.get("session").flash("flashMessages", {
        error: ["error", "error2"],
        info: ["info"],
        success: ["success"],
        warning: ["warning"],
      });

      return c.render("");
    });

    app.get("/test/error", (c) => {
      const { type } = c.req.query();

      if (type === "http") {
        throw new HTTPException(400, {
          res: new Response(null, { headers: { "x-foo": "bar" } }),
        });
      }

      if (type === "http422") {
        throw new HTTPException(422, {
          message: "foo",
          cause: { foo: ["yes"] },
        });
      }

      if (type === "valierr") {
        const e = new errors.E_VALIDATION_ERROR([{
          message: "lorem ipsum",
          field: "foo",
        }, {
          message: "lorem lorem",
          field: "bar",
        }]);

        e.message = "foo";

        throw e;
      }

      if (type === "multparterr") {
        throw new MultipartParseError("foo");
      }

      throw new Error("foo");
    });
  }

  return app.route("/", router());
};
