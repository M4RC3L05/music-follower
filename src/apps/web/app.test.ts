import {
  afterAll,
  beforeAll,
  beforeEach,
  describe,
  it,
} from "@std/testing/bdd";
import { type CustomDatabase, makeDatabase } from "#src/database/mod.ts";
import { testDbUtils, testUtils } from "#src/common/utils/mod.ts";
import { makeApp } from "#src/apps/web/app.tsx";
import type { Hono } from "@hono/hono";
import { assertEquals, assertNotEquals } from "@std/assert";
import { MemoryStore } from "@jcs224/hono-sessions";
import * as testFixtures from "#src/common/test-fixtures/mod.ts";

let app: Hono;
let db: CustomDatabase;
let memStore: MemoryStore;

beforeAll(async () => {
  db = makeDatabase();
  await testDbUtils.runMigrations(db);

  memStore = new MemoryStore();
  app = makeApp({ database: db, sessioStore: memStore });
});

beforeEach(() => {
  db.exec("delete from accounts");
  testFixtures.loadAccount(db);
  // deno-lint-ignore ban-ts-comment
  // @ts-ignore
  memStore.data.clear();
});

afterAll(() => {
  db.close();
});

describe("App", () => {
  describe("snapshots", () => {
    it("should render default layout correctly", async (t) => {
      const res = await app.request("/test/flash-messages");
      const dom = await testUtils.getDom(res);

      assertEquals(res.status, 200);
      await testUtils.expectHtmlSnapshot(t, dom);
    });

    it("should render default layout correctly loggedin", async (t) => {
      const auth = await testUtils.authenticateRequest(app);
      const res = await app.request("/test/flash-messages", {
        headers: { cookie: auth.sid },
      });
      const dom = await testUtils.getDom(res);

      assertEquals(res.status, 200);
      await testUtils.expectHtmlSnapshot(t, dom);
    });
  });

  it("should redirect to login page if not authenticated", async () => {
    const response = await app.request("/");

    assertEquals(response.status, 302);
    assertEquals(response.headers.get("location"), "/auth/login");
  });

  it("should render a not found error page if route what not found", async () => {
    const response = await app.request("/not/found");
    const dom = await testUtils.getDom(response);

    assertEquals(response.status, 404);
    testUtils.assertNodeExists(dom, "main");
    testUtils.assertNodeNotExists(dom, "nav.navbar");
    testUtils.assertSeeInOrder(dom, [
      "<h1>Not found</h1>",
      '<a href="/">Go Home</a>',
    ], "body main");
  });

  it("should disable caching for not static routes", async () => {
    const response = await app.request("/test/index");

    assertEquals(response.status, 200);
    assertEquals(
      response.headers.get("cache-control"),
      "no-cache, no-store, must-revalidate",
    );
    assertEquals(response.headers.get("pragma"), "no-cache");
    assertEquals(response.headers.get("expires"), "0");

    const response2 = await app.request("/public/foo");

    assertNotEquals(
      response2.headers.get("cache-control"),
      "no-cache, no-store, must-revalidate",
    );
    assertNotEquals(response2.headers.get("pragma"), "no-cache");
    assertNotEquals(response2.headers.get("expires"), "0");
  });

  it("should check againt csrf", async () => {
    const response = await app.request("/test/index", {
      method: "post",
      body: new URLSearchParams({ "foo": "bar" }),
    });

    assertEquals(response.status, 403);

    const response2 = await app.request("/test/index", {
      method: "post",
      headers: {
        origin: "http://localhosts",
      },
      body: new URLSearchParams({ "foo": "bar" }),
    });

    assertEquals(response2.status, 403);

    const response3 = await app.request("/test/index", {
      method: "post",
      headers: {
        origin: "http://localhost",
      },
      body: new URLSearchParams({ "foo": "bar" }),
    });

    assertEquals(response3.status, 404);
  });

  it("should render flash messages", async (t) => {
    const res = await app.request("/test/flash-messages");
    const dom = await testUtils.getDom(res);

    assertEquals(res.status, 200);
    await testUtils.expectHtmlSnapshot(t, dom, "#flash-messages");
  });

  it("should catch unknown errors and return 500", async (t) => {
    const res = await app.request("/test/error");
    const dom = await testUtils.getDom(res);

    assertEquals(res.status, 500);
    await testUtils.expectHtmlSnapshot(t, dom, "body");
  });

  it("should catch http errors and return their status code and proxy any headers", async (t) => {
    const res = await app.request("/test/error?type=http");
    const dom = await testUtils.getDom(res);

    assertEquals(res.status, 400);
    assertEquals(res.headers.get("x-foo"), "bar");
    await testUtils.expectHtmlSnapshot(t, dom, "body");
  });

  it("should catch http 422 errors and return their status code and refirect to referer is exists", async () => {
    {
      const res = await app.request("/test/error?type=http422");

      assertEquals(res.status, 302);
      assertEquals(res.headers.get("location"), "/");

      const res2 = await app.request("/test/index", {
        headers: { cookie: res.headers.get("set-cookie")! },
      });
      const dom2 = await testUtils.getDom(res2);
      assertEquals(res2.status, 200);
      testUtils.assertNodeExists(dom2, "body #flash-messages");
      testUtils.assertSeeText(dom2, "foo", "body #flash-messages");
    }

    {
      const res = await app.request("/test/error?type=http422", {
        headers: { "referer": "http://localhost/foo" },
      });

      assertEquals(res.status, 302);
      assertEquals(res.headers.get("location"), "http://localhost/foo");

      const res2 = await app.request("/test/index", {
        headers: { cookie: res.headers.get("set-cookie")! },
      });
      const dom2 = await testUtils.getDom(res2);
      assertEquals(res2.status, 200);
      testUtils.assertNodeExists(dom2, "body #flash-messages");
      testUtils.assertSeeText(dom2, "foo", "body #flash-messages");
    }
  });

  it("should catch validation errors and return 422 status code and refirect to referer is exists", async () => {
    {
      const res = await app.request("/test/error?type=valierr");

      const res2 = await app.request("/test/index", {
        headers: { cookie: res.headers.get("set-cookie")! },
      });
      const dom2 = await testUtils.getDom(res2);
      assertEquals(res2.status, 200);
      testUtils.assertNodeExists(dom2, "body #flash-messages");
      testUtils.assertSeeText(dom2, "foo", "body #flash-messages");
    }

    {
      const res = await app.request("/test/error?type=valierr", {
        headers: { "referer": "http://localhost/foo" },
      });

      assertEquals(res.status, 302);
      assertEquals(res.headers.get("location"), "http://localhost/foo");

      const res2 = await app.request("/test/index", {
        headers: { cookie: res.headers.get("set-cookie")! },
      });
      const dom2 = await testUtils.getDom(res2);
      assertEquals(res2.status, 200);
      testUtils.assertNodeExists(dom2, "body #flash-messages");
      testUtils.assertSeeText(dom2, "foo", "body #flash-messages");
    }
  });

  it("should catch MultipartParseError errors and return 422 status code and refirect to referer is exists", async () => {
    {
      const auth = await testUtils.authenticateRequest(app);
      const res = await app.request("/test/error?type=valierr", {
        headers: { cookie: auth.sid },
      });

      assertEquals(res.status, 302);
      assertEquals(res.headers.get("location"), "/");

      const res2 = await app.request("/test/index", {
        headers: { cookie: auth.sid },
      });
      const dom2 = await testUtils.getDom(res2);
      assertEquals(res2.status, 200);
      testUtils.assertNodeExists(dom2, "body #flash-messages");
      testUtils.assertSeeText(dom2, "foo", "body #flash-messages");
    }

    {
      const res = await app.request("/test/error?type=valierr", {
        headers: { "referer": "http://localhost/foo" },
      });

      assertEquals(res.status, 302);
      assertEquals(res.headers.get("location"), "http://localhost/foo");

      const res2 = await app.request("/test/index", {
        headers: { cookie: res.headers.get("set-cookie")! },
      });
      const dom2 = await testUtils.getDom(res2);
      testUtils.assertNodeExists(dom2, "body #flash-messages");
      testUtils.assertSeeText(dom2, "foo", "body #flash-messages");
    }
  });
});
