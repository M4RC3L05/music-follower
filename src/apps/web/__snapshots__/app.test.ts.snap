export const snapshot = {};

snapshot[`global > App > snapshots > should render default layout correctly 1`] = `
'
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta
      http-equiv="X-UA-Compatible"
      content="IE=edge"
    >
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0"
    >
    <title>
      MUSIC FOLLOWER | ADMIN
    </title>
    <link
      rel="stylesheet"
      href="/public/css/main.css"
    >
    <script type="module">
      const getPreferredTheme = () => {
              return globalThis.matchMedia("(prefers-color-scheme: dark)").matches
                ? "dark"
                : "light";
            };

            const setTheme = (theme) => {
              document.documentElement.setAttribute("data-bs-theme", theme);
            };

            setTheme(getPreferredTheme());

            globalThis.matchMedia("(prefers-color-scheme: dark)").addEventListener(
              "change",
              () => {
                setTheme(getPreferredTheme());
              },
            );
    </script>
  </head>
  <body>
    <nav class="navbar navbar-expand-lg bg-body-tertiary">
      <div class="container">
        <a
          class="navbar-brand"
          href="/auth/login"
        >
          Music Follower
        </a>
        <button
          class="navbar-toggler"
          type="button"
          data-bs-toggle="collapse"
          data-bs-target="#navbarSupportedContent"
          aria-controls="navbarSupportedContent"
          aria-expanded="false"
          aria-label="Toggle navigation"
        >
          <span class="navbar-toggler-icon">
          </span>
        </button>
        <div
          class="collapse navbar-collapse"
          id="navbarSupportedContent"
        >
          <ul class="navbar-nav me-auto mb-2 mb-lg-0">
            <li class="nav-item">
              <a
                class="nav-link "
                href="/auth/login"
              >
                Login
              </a>
            </li>
            <li class="nav-item">
              <a
                class="nav-link "
                href="/auth/register"
              >
                Register
              </a>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <div
      id="flash-messages"
      class="position-fixed z-3 vstack gap-2 overflow-x-hidden overflow-y-scroll"
      style="bottom:1.3rem;right:1.3rem;width:600px;top:1.3rem;pointer-events:none"
    >
      <div
        class="alert alert-danger alert-dismissible fade show mb-0 d-flex align-items-center"
        role="alert"
        style="pointer-events:all"
      >
        <i class="bi bi-x-circle me-2">
        </i>
        error
        <button
          type="button"
          class="btn-close"
          data-bs-dismiss="alert"
          aria-label="Close"
        >
        </button>
      </div>
      <div
        class="alert alert-danger alert-dismissible fade show mb-0 d-flex align-items-center"
        role="alert"
        style="pointer-events:all"
      >
        <i class="bi bi-x-circle me-2">
        </i>
        error2
        <button
          type="button"
          class="btn-close"
          data-bs-dismiss="alert"
          aria-label="Close"
        >
        </button>
      </div>
      <div
        class="alert alert-warning alert-dismissible fade show mb-0 d-flex align-items-center"
        role="alert"
        style="pointer-events:all"
      >
        <i class="bi bi-exclamation-triangle me-2">
        </i>
        warning
        <button
          type="button"
          class="btn-close"
          data-bs-dismiss="alert"
          aria-label="Close"
        >
        </button>
      </div>
      <div
        class="alert alert-success alert-dismissible fade show mb-0 d-flex align-items-center"
        role="alert"
        style="pointer-events:all"
      >
        <i class="bi bi-check-square me-2">
        </i>
        success
        <button
          type="button"
          class="btn-close"
          data-bs-dismiss="alert"
          aria-label="Close"
        >
        </button>
      </div>
      <div
        class="alert alert-info alert-dismissible fade show mb-0 d-flex align-items-center"
        role="alert"
        style="pointer-events:all"
      >
        <i class="bi bi-exclamation-square me-2">
        </i>
        info
        <button
          type="button"
          class="btn-close"
          data-bs-dismiss="alert"
          aria-label="Close"
        >
        </button>
      </div>
    </div>
    <script
      type="module"
      src="/deps/bootstrap/dist/js/bootstrap.bundle.js"
    >
    </script>
  </body>
</html>
'
`;

snapshot[`global > App > snapshots > should render default layout correctly loggedin 1`] = `
'
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta
      http-equiv="X-UA-Compatible"
      content="IE=edge"
    >
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0"
    >
    <title>
      MUSIC FOLLOWER | ADMIN
    </title>
    <link
      rel="stylesheet"
      href="/public/css/main.css"
    >
    <script type="module">
      const getPreferredTheme = () => {
              return globalThis.matchMedia("(prefers-color-scheme: dark)").matches
                ? "dark"
                : "light";
            };

            const setTheme = (theme) => {
              document.documentElement.setAttribute("data-bs-theme", theme);
            };

            setTheme(getPreferredTheme());

            globalThis.matchMedia("(prefers-color-scheme: dark)").addEventListener(
              "change",
              () => {
                setTheme(getPreferredTheme());
              },
            );
    </script>
  </head>
  <body>
    <nav class="navbar navbar-expand-lg bg-body-tertiary">
      <div class="container">
        <a
          class="navbar-brand"
          href="/"
        >
          Music Follower
        </a>
        <button
          class="navbar-toggler"
          type="button"
          data-bs-toggle="collapse"
          data-bs-target="#navbarSupportedContent"
          aria-controls="navbarSupportedContent"
          aria-expanded="false"
          aria-label="Toggle navigation"
        >
          <span class="navbar-toggler-icon">
          </span>
        </button>
        <div
          class="collapse navbar-collapse"
          id="navbarSupportedContent"
        >
          <ul class="navbar-nav me-auto mb-2 mb-lg-0">
            <li class="nav-item">
              <a
                class="nav-link "
                aria-current="page"
                href="/"
              >
                Home
              </a>
            </li>
            <li class="nav-item">
              <a
                class="nav-link "
                href="/artists"
              >
                Artists
              </a>
            </li>
            <li class="nav-item">
              <a
                class="nav-link "
                href="/releases"
              >
                Releases
              </a>
            </li>
            <div
              class="modal fade"
              id="confirm-delete-account-modal"
              tabindex="-1"
            >
              <div class="modal-dialog">
                <div class="modal-content">
                  <div class="modal-header">
                    <h5 class="modal-title">
                      Confirm delete account
                    </h5>
                    <button
                      type="button"
                      class="btn-close"
                      data-bs-dismiss="modal"
                      aria-label="Close"
                    >
                    </button>
                  </div>
                  <div class="modal-body">
                    <p>
                      Are you sure you want to delet the current auth account?
                    </p>
                    <p>
                      You will be able to create a new one.
                    </p>
                  </div>
                  <div class="modal-footer">
                    <form
                      method="post"
                      action="/auth/delete"
                    >
                      <input
                        class="btn btn-outline-primary"
                        type="submit"
                        value="Confirm"
                      >
                    </form>
                    <button
                      type="button"
                      class="btn btn-outline-secondary"
                      data-bs-dismiss="modal"
                    >
                      Cancel
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <li
              class="nav-item"
              data-bs-toggle="modal"
              data-bs-target="#confirm-delete-account-modal"
            >
              <a
                class="nav-link"
                style="cursor:pointer"
              >
                Delete account?
              </a>
            </li>
            <li class="nav-item">
              <form
                method="post"
                action="/auth/logout"
              >
                <input
                  type="submit"
                  value="Logout"
                  class="nav-link"
                >
              </form>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <div
      id="flash-messages"
      class="position-fixed z-3 vstack gap-2 overflow-x-hidden overflow-y-scroll"
      style="bottom:1.3rem;right:1.3rem;width:600px;top:1.3rem;pointer-events:none"
    >
      <div
        class="alert alert-danger alert-dismissible fade show mb-0 d-flex align-items-center"
        role="alert"
        style="pointer-events:all"
      >
        <i class="bi bi-x-circle me-2">
        </i>
        error
        <button
          type="button"
          class="btn-close"
          data-bs-dismiss="alert"
          aria-label="Close"
        >
        </button>
      </div>
      <div
        class="alert alert-danger alert-dismissible fade show mb-0 d-flex align-items-center"
        role="alert"
        style="pointer-events:all"
      >
        <i class="bi bi-x-circle me-2">
        </i>
        error2
        <button
          type="button"
          class="btn-close"
          data-bs-dismiss="alert"
          aria-label="Close"
        >
        </button>
      </div>
      <div
        class="alert alert-warning alert-dismissible fade show mb-0 d-flex align-items-center"
        role="alert"
        style="pointer-events:all"
      >
        <i class="bi bi-exclamation-triangle me-2">
        </i>
        warning
        <button
          type="button"
          class="btn-close"
          data-bs-dismiss="alert"
          aria-label="Close"
        >
        </button>
      </div>
      <div
        class="alert alert-success alert-dismissible fade show mb-0 d-flex align-items-center"
        role="alert"
        style="pointer-events:all"
      >
        <i class="bi bi-check-square me-2">
        </i>
        success
        <button
          type="button"
          class="btn-close"
          data-bs-dismiss="alert"
          aria-label="Close"
        >
        </button>
      </div>
      <div
        class="alert alert-info alert-dismissible fade show mb-0 d-flex align-items-center"
        role="alert"
        style="pointer-events:all"
      >
        <i class="bi bi-exclamation-square me-2">
        </i>
        info
        <button
          type="button"
          class="btn-close"
          data-bs-dismiss="alert"
          aria-label="Close"
        >
        </button>
      </div>
    </div>
    <script
      type="module"
      src="/deps/bootstrap/dist/js/bootstrap.bundle.js"
    >
    </script>
  </body>
</html>
'
`;

snapshot[`global > App > should render flash messages 1`] = `
'
<div
  id="flash-messages"
  class="position-fixed z-3 vstack gap-2 overflow-x-hidden overflow-y-scroll"
  style="bottom:1.3rem;right:1.3rem;width:600px;top:1.3rem;pointer-events:none"
>
  <div
    class="alert alert-danger alert-dismissible fade show mb-0 d-flex align-items-center"
    role="alert"
    style="pointer-events:all"
  >
    <i class="bi bi-x-circle me-2">
    </i>
    error
    <button
      type="button"
      class="btn-close"
      data-bs-dismiss="alert"
      aria-label="Close"
    >
    </button>
  </div>
  <div
    class="alert alert-danger alert-dismissible fade show mb-0 d-flex align-items-center"
    role="alert"
    style="pointer-events:all"
  >
    <i class="bi bi-x-circle me-2">
    </i>
    error2
    <button
      type="button"
      class="btn-close"
      data-bs-dismiss="alert"
      aria-label="Close"
    >
    </button>
  </div>
  <div
    class="alert alert-warning alert-dismissible fade show mb-0 d-flex align-items-center"
    role="alert"
    style="pointer-events:all"
  >
    <i class="bi bi-exclamation-triangle me-2">
    </i>
    warning
    <button
      type="button"
      class="btn-close"
      data-bs-dismiss="alert"
      aria-label="Close"
    >
    </button>
  </div>
  <div
    class="alert alert-success alert-dismissible fade show mb-0 d-flex align-items-center"
    role="alert"
    style="pointer-events:all"
  >
    <i class="bi bi-check-square me-2">
    </i>
    success
    <button
      type="button"
      class="btn-close"
      data-bs-dismiss="alert"
      aria-label="Close"
    >
    </button>
  </div>
  <div
    class="alert alert-info alert-dismissible fade show mb-0 d-flex align-items-center"
    role="alert"
    style="pointer-events:all"
  >
    <i class="bi bi-exclamation-square me-2">
    </i>
    info
    <button
      type="button"
      class="btn-close"
      data-bs-dismiss="alert"
      aria-label="Close"
    >
    </button>
  </div>
</div>
'
`;

snapshot[`global > App > should catch unknown errors and return 500 1`] = `
'
<body>
  <main>
    <div class="container mt-5">
      <div class="row">
        <div class="col text-center">
          <h1>
            Something broke...
          </h1>
          <a href="/">
            Go Home
          </a>
        </div>
      </div>
    </div>
  </main>
  <script
    type="module"
    src="/deps/bootstrap/dist/js/bootstrap.bundle.js"
  >
  </script>
</body>
'
`;

snapshot[`global > App > should catch http errors and return their status code and proxy any headers 1`] = `
'
<body>
  <main>
    <div class="container mt-5">
      <div class="row">
        <div class="col text-center">
          <h1>
            Bad Request
          </h1>
          <a href="/">
            Go Home
          </a>
        </div>
      </div>
    </div>
  </main>
  <script
    type="module"
    src="/deps/bootstrap/dist/js/bootstrap.bundle.js"
  >
  </script>
</body>
'
`;
